crates/sui-graphql-rpc/src/types/checkpoint.rs

```diff
@@ -15,8 +19,52 @@ pub(crate) struct Checkpoint {
     pub live_object_set_digest: Option<String>,
     pub network_total_transactions: Option<u64>,
     pub rolling_gas_summary: Option<GasCostSummary>,
-    pub epoch: Option<Epoch>,
+    pub epoch: Epoch,
```

Comment:
I would have thought you wanted to do something similar to your other refactoring, i.e. make `Checkpoint` an `Object` and fetch its epoch on-demand?

Reply:
Yeah I guess I haven't fully convinced myself on what standard to follow. Implemented as so in the PR, even though the Epoch field is instantiated always, its expensive fields protocol_config and sui system state summary related fields are still deferred. If we were to port the epoch field out, it'd look like the below which I consider to be a data conversion. Although I should change the typing back to Option<Epoch> to conform to the schema

```
#[ComplexObject]
impl Epoch {
    async fn epoch(&self) -> Epoch {
        Epoch {
            epoch_id: self.epoch_id,
            gas_cost_summary: self.gas_cost_summary.clone(),
        }
    }
}
```

Reply:
I see -- I think it's important for consistency to have a convention that we stick with.  Thinking forward to when these queries will be served by database accesses, where we can select precise fields, rather than RPC queries where we get all the data up-front, I think the model should be something like this:

- For types that correspond to tables in the data layer (transactions, objects, events, epochs, checkpoints, etc), the Rust representation in GraphQL should contain just the primary key(s).  This data should always be "free" to serve because even when you're requesting primary key fields for a type that is nested in another type, you would not need to do an extra fetch to get them, because these are the fields that one type uses to refer to another type (i.e. they are also foreign keys).
- For types that correspond to serialized data (move types, move values, transaction kinds, signatures, etc) the Rust representation can contain everything that is trivially deserialized from the stored representation.

How does that sound @wlmyng ?

Reply:
Fantastic, that makes a great deal of sense to me :fire: